diff --git a/Cargo.lock b/Cargo.lock
index 0a3feac2e..3cb3e74d1 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -483,7 +483,7 @@ version = "4.0.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "85a8ab73a1c02b0c15597b22e09c7dc36e63b2f601f9d1e83ac0c3decd38b1ae"
 dependencies = [
- "nix 0.29.0",
+ "nix",
  "terminfo",
  "thiserror 2.0.12",
  "which",
@@ -501,7 +501,7 @@ dependencies = [
  "colored 2.2.0",
  "glob",
  "libc",
- "nix 0.29.0",
+ "nix",
  "serde",
  "serde_json",
  "statrs",
@@ -824,16 +824,6 @@ dependencies = [
  "typenum",
 ]
 
-[[package]]
-name = "ctrlc"
-version = "3.4.7"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "46f93780a459b7d656ef7f071fe699c4d3d2cb201c4b24d085b6ddc505276e73"
-dependencies = [
- "nix 0.30.1",
- "windows-sys 0.59.0",
-]
-
 [[package]]
 name = "darling"
 version = "0.20.11"
@@ -2020,18 +2010,6 @@ dependencies = [
  "libc",
 ]
 
-[[package]]
-name = "nix"
-version = "0.30.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "74523f3a35e05aba87a1d978330aef40f67b0304ac79c1c00b294c9830543db6"
-dependencies = [
- "bitflags 2.9.1",
- "cfg-if",
- "cfg_aliases",
- "libc",
-]
-
 [[package]]
 name = "nom"
 version = "7.1.3"
@@ -4197,7 +4175,6 @@ dependencies = [
  "clearscreen",
  "colored 3.0.0",
  "crossbeam",
- "ctrlc",
  "dunce",
  "filetime",
  "indicatif",
diff --git a/crates/ruff/src/cache.rs b/crates/ruff/src/cache.rs
index d782d8c9e..cf20696b3 100644
--- a/crates/ruff/src/cache.rs
+++ b/crates/ruff/src/cache.rs
@@ -57,6 +57,8 @@ impl FileCacheKey {
         };
         #[cfg(windows)]
         let permissions: u32 = metadata.permissions().readonly().into();
+        #[cfg(not(any(unix, windows)))]
+        let permissions: u32 = 0644u32;
 
         Ok(FileCacheKey {
             file_last_modified: FileTime::from_last_modification_time(&metadata),
diff --git a/crates/ruff/src/commands/add_noqa.rs b/crates/ruff/src/commands/add_noqa.rs
index d5eaeb017..e39eee731 100644
--- a/crates/ruff/src/commands/add_noqa.rs
+++ b/crates/ruff/src/commands/add_noqa.rs
@@ -44,7 +44,7 @@ pub(crate) fn add_noqa(
 
     let start = Instant::now();
     let modifications: usize = paths
-        .par_iter()
+        .iter()
         .flatten()
         .filter_map(|resolved_file| {
             let SourceType::Python(source_type @ (PySourceType::Python | PySourceType::Stub)) =
diff --git a/crates/ruff/src/commands/check.rs b/crates/ruff/src/commands/check.rs
index 0c0ed5995..e3f23831a 100644
--- a/crates/ruff/src/commands/check.rs
+++ b/crates/ruff/src/commands/check.rs
@@ -65,7 +65,7 @@ pub(crate) fn check(
     };
 
     let start = Instant::now();
-    let diagnostics_per_file = paths.par_iter().filter_map(|resolved_file| {
+    let diagnostics_per_file = paths.iter().filter_map(|resolved_file| {
         let result = match resolved_file {
             Ok(resolved_file) => {
                 let path = resolved_file.path();
@@ -150,17 +150,12 @@ pub(crate) fn check(
 
     // Aggregate the diagnostics of all checked files and count the checked files.
     // This can't be a regular for loop because we use `par_iter`.
-    let (mut all_diagnostics, checked_files) = diagnostics_per_file
-        .fold(
-            || (Diagnostics::default(), 0u64),
-            |(all_diagnostics, checked_files), file_diagnostics| {
-                (all_diagnostics + file_diagnostics, checked_files + 1)
-            },
-        )
-        .reduce(
-            || (Diagnostics::default(), 0u64),
-            |a, b| (a.0 + b.0, a.1 + b.1),
-        );
+    let (mut all_diagnostics, checked_files) = diagnostics_per_file.fold(
+        (Diagnostics::default(), 0u64),
+        |(all_diagnostics, checked_files), file_diagnostics| {
+            (all_diagnostics + file_diagnostics, checked_files + 1)
+        },
+    );
 
     all_diagnostics
         .inner
diff --git a/crates/ty/Cargo.toml b/crates/ty/Cargo.toml
index 4189758bb..fee054557 100644
--- a/crates/ty/Cargo.toml
+++ b/crates/ty/Cargo.toml
@@ -29,7 +29,7 @@ clap_complete_command = { workspace = true }
 clearscreen = { workspace = true }
 colored = { workspace = true }
 crossbeam = { workspace = true }
-ctrlc = { version = "3.4.4" }
+#ctrlc = { version = "3.4.4" }
 indicatif = { workspace = true }
 jiff = { workspace = true }
 rayon = { workspace = true }
diff --git a/crates/ty/src/lib.rs b/crates/ty/src/lib.rs
index f16192bcd..3a78e9849 100644
--- a/crates/ty/src/lib.rs
+++ b/crates/ty/src/lib.rs
@@ -138,6 +138,7 @@ fn run_check(args: CheckCommand) -> anyhow::Result<ExitStatus> {
 
     // Listen to Ctrl+C and abort the watch mode.
     let main_loop_cancellation_token = Mutex::new(Some(main_loop_cancellation_token));
+    /*
     ctrlc::set_handler(move || {
         let mut lock = main_loop_cancellation_token.lock().unwrap();
 
@@ -145,6 +146,7 @@ fn run_check(args: CheckCommand) -> anyhow::Result<ExitStatus> {
             token.stop();
         }
     })?;
+    */
 
     let exit_status = if watch {
         main_loop.watch(&mut db)?

