diff --git a/Cargo.toml b/Cargo.toml
index bed513fe3..f653a70a6 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -293,3 +293,7 @@ debug = 1
 # The profile that 'cargo dist' will build with.
 [profile.dist]
 inherits = "release"
+
+[patch.crates-io]
+ignore = { path = "../ripgrep/crates/ignore" }
+filetime = { path = "../filetime" }
diff --git a/crates/ruff/src/cache.rs b/crates/ruff/src/cache.rs
index 6b696b55d..3d9bf9f4b 100644
--- a/crates/ruff/src/cache.rs
+++ b/crates/ruff/src/cache.rs
@@ -48,9 +48,11 @@ impl FileCacheKey {
         };
         #[cfg(windows)]
         let permissions: u32 = metadata.permissions().readonly().into();
+        #[cfg(not(any(unix, windows)))]
+        let permissions: u32 = 0755u32;
 
         Ok(FileCacheKey {
-            file_last_modified: FileTime::from_last_modification_time(&metadata),
+            file_last_modified: FileTime::zero(),
             file_permissions_mode: permissions,
         })
     }
diff --git a/crates/ruff/src/commands/add_noqa.rs b/crates/ruff/src/commands/add_noqa.rs
index d5eaeb017..e39eee731 100644
--- a/crates/ruff/src/commands/add_noqa.rs
+++ b/crates/ruff/src/commands/add_noqa.rs
@@ -44,7 +44,7 @@ pub(crate) fn add_noqa(
 
     let start = Instant::now();
     let modifications: usize = paths
-        .par_iter()
+        .iter()
         .flatten()
         .filter_map(|resolved_file| {
             let SourceType::Python(source_type @ (PySourceType::Python | PySourceType::Stub)) =
diff --git a/crates/ruff/src/commands/check.rs b/crates/ruff/src/commands/check.rs
index 0c0ed5995..e3f23831a 100644
--- a/crates/ruff/src/commands/check.rs
+++ b/crates/ruff/src/commands/check.rs
@@ -65,7 +65,7 @@ pub(crate) fn check(
     };
 
     let start = Instant::now();
-    let diagnostics_per_file = paths.par_iter().filter_map(|resolved_file| {
+    let diagnostics_per_file = paths.iter().filter_map(|resolved_file| {
         let result = match resolved_file {
             Ok(resolved_file) => {
                 let path = resolved_file.path();
@@ -150,17 +150,12 @@ pub(crate) fn check(
 
     // Aggregate the diagnostics of all checked files and count the checked files.
     // This can't be a regular for loop because we use `par_iter`.
-    let (mut all_diagnostics, checked_files) = diagnostics_per_file
-        .fold(
-            || (Diagnostics::default(), 0u64),
-            |(all_diagnostics, checked_files), file_diagnostics| {
-                (all_diagnostics + file_diagnostics, checked_files + 1)
-            },
-        )
-        .reduce(
-            || (Diagnostics::default(), 0u64),
-            |a, b| (a.0 + b.0, a.1 + b.1),
-        );
+    let (mut all_diagnostics, checked_files) = diagnostics_per_file.fold(
+        (Diagnostics::default(), 0u64),
+        |(all_diagnostics, checked_files), file_diagnostics| {
+            (all_diagnostics + file_diagnostics, checked_files + 1)
+        },
+    );
 
     all_diagnostics
         .inner
diff --git a/crates/ruff_workspace/src/resolver.rs b/crates/ruff_workspace/src/resolver.rs
index 84a31ebbe..f7c1aef6e 100644
--- a/crates/ruff_workspace/src/resolver.rs
+++ b/crates/ruff_workspace/src/resolver.rs
@@ -486,12 +486,6 @@ pub fn python_files_in_path<'a>(
     builder.standard_filters(resolver.respect_gitignore());
     builder.hidden(false);
 
-    builder.threads(
-        std::thread::available_parallelism()
-            .map_or(1, std::num::NonZeroUsize::get)
-            .min(12),
-    );
-
     let walker = builder.build_parallel();
 
     // Run the `WalkParallel` to collect all Python files.
diff --git a/crates/ty/Cargo.toml b/crates/ty/Cargo.toml
index 4189758bb..fee054557 100644
--- a/crates/ty/Cargo.toml
+++ b/crates/ty/Cargo.toml
@@ -29,7 +29,7 @@ clap_complete_command = { workspace = true }
 clearscreen = { workspace = true }
 colored = { workspace = true }
 crossbeam = { workspace = true }
-ctrlc = { version = "3.4.4" }
+#ctrlc = { version = "3.4.4" }
 indicatif = { workspace = true }
 jiff = { workspace = true }
 rayon = { workspace = true }
diff --git a/crates/ty/src/lib.rs b/crates/ty/src/lib.rs
index f16192bcd..3a78e9849 100644
--- a/crates/ty/src/lib.rs
+++ b/crates/ty/src/lib.rs
@@ -138,6 +138,7 @@ fn run_check(args: CheckCommand) -> anyhow::Result<ExitStatus> {
 
     // Listen to Ctrl+C and abort the watch mode.
     let main_loop_cancellation_token = Mutex::new(Some(main_loop_cancellation_token));
+    /*
     ctrlc::set_handler(move || {
         let mut lock = main_loop_cancellation_token.lock().unwrap();
 
@@ -145,6 +146,7 @@ fn run_check(args: CheckCommand) -> anyhow::Result<ExitStatus> {
             token.stop();
         }
     })?;
+    */
 
     let exit_status = if watch {
         main_loop.watch(&mut db)?
